src/irmin/store_intf.ml

```diff
@@ -122,6 +122,33 @@ module type S = sig
     val import : t -> slice -> (unit, [ `Msg of string ]) result Lwt.t
     (** [import t s] imports the contents of the slice [s] in [t]. Does not
         modify branches. *)
+
+    val iter_nodes :
+      t ->
+      min:hash list ->
+      max:hash list ->
+      node:(hash -> unit Lwt.t) ->
+      contents:(hash * metadata -> unit Lwt.t) ->
+      skip:(hash -> bool Lwt.t) ->
+      unit ->
+      unit Lwt.t
+    (** [iter t ~min ~max ~node ~contents ~skip ()] iterates in reverse
+        topological order over the closure graph of [t] starting with the [max]
+        nodes and bounded by the [min] nodes, as specified by
+        {{!Irmin__.S.NODE_GRAPH.iter} GRAPH.iter}. *)
+
+    val iter_commits :
```

Comment:
I have a feeling that this function is _also_ going to iterate over the nodes and the contents referenced by the closure of commits given, it's just not going to do anything when it visits them; is that right?

Reply:
Yes, it just ignores them. So we can add `?node:(node -> unit Lwt.t)` and `?contents:(contents * metadata -> unit Lwt.t)`. I didn't do it because I don't need it in the layered store, but maybe we should add them for future usage?

Reply:
Not having iterators for `node` and `contents` seems fine. I was mostly concerned about performance: iterating over all objects in the closure when you're only making progress for the commits feels like it might be quite inefficient.

Reply:
Good point, it actually only iterates over commits, as the predecessors of a commit are [its parents](https://github.com/mirage/irmin/pull/1077/files#diff-be270f772293fb2b19da1931f6aafcd1R200-R202). Unlike for nodes where predecessors can be both nodes and contents. So I was wrong above, we cannot iterate over nodes and contents with this function.

Reply:
Hmm... indeed. This `iter` function never seems to behave the way I expect it too :stuck_out_tongue:

Can you add something like `` `Node _ | `Contents _ -> assert false `` in the implementation, just to make that clear?

Reply:
Nice suggestions, thanks!